keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/GitDataGenerator.kt

```diff
@@ -0,0 +1,65 @@
+package com.netflix.spinnaker.keel.slack.handlers
+
+import com.netflix.spinnaker.keel.api.ScmInfo
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.api.artifacts.PublishedArtifact
+import com.netflix.spinnaker.keel.artifacts.getScmBaseLink
+import com.slack.api.model.kotlin_extension.block.SectionBlockBuilder
+import org.apache.logging.log4j.util.Strings
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+
+@Component
+class GitDataGenerator(
+  private val scmInfo: ScmInfo,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+) {
+
+  fun generateRepoLink(gitMetadata: GitMetadata?): String {
+    if (gitMetadata != null) {
+      val baseScmUrl = gitMetadata.commitInfo?.link?.let { getScmBaseLink(scmInfo, it) }
+      return "$baseScmUrl/projects/${gitMetadata.project}/repos/${gitMetadata.repo?.name}"
+    } else return Strings.EMPTY
+  }
+
+  /**
+   * generateGitData will create a slack section blocks, which looks like:
+   * "spkr/keel › PR#7 › master › c25a357"
+   * Or: "spkr/keel › master › c25a358" (no PR data)
+   * Each component will have the corresponding link attached to SCM
+   */
+  fun generateData(sectionBlockBuilder: SectionBlockBuilder, application: String, artifact: PublishedArtifact): SectionBlockBuilder {
+//  protected fun SectionBlockBuilder.generateGitData(application: String, artifact: PublishedArtifact) {
+    with(sectionBlockBuilder) {
+      var details = ""
+      val artifactUrl = "$spinnakerBaseUrl/#/applications/${application}/environments/${artifact.reference}/${artifact.version}"
+
+      if (artifact.gitMetadata != null) {
+        val repoLink = generateRepoLink(artifact.gitMetadata)
+        with(artifact.gitMetadata) {
+
+          details += "<$repoLink|${this?.project}/" +
+            "${this?.repo?.name}> › " +
```

Comment:
I can't tell what exactly will end up in the string if any of these things are null. Are you sure this will render as intended in those cases?

Reply:
added more null checks on top.